File: 215.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        int lo =  0, hi =  nums.size()-1;
        while (1)
        {
            int mid = paration(nums, lo, hi);
            if (mid == k-1) return nums[mid];
            else if (mid > k-1) {
                hi = mid - 1;
            }
            else lo = mid + 1;
        }
        return nums[k-1];
    }

    int paration(vector<int>& nums, int lo, int hi) {
        int pivot = nums[lo];
        while (lo < hi) {
            while (lo < hi && pivot >= nums[hi]) {
                --hi;
            }
            nums[lo] = nums[hi];
            while (lo < hi && pivot <= nums[lo]) {
                ++lo;
            }
            nums[hi] = nums[lo];
        }
        nums[lo] = pivot;
        return lo;
    }
};

File: 215.py
from typing import List
from random import randint


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        def paration(lo, hi):
            rand_idx = randint(lo, hi)
            nums[lo], nums[rand_idx] = nums[rand_idx], nums[lo]
            pivot = nums[lo]
            while lo < hi:
                while lo < hi and nums[hi] <= pivot:
                    hi -= 1
                nums[lo] = nums[hi]
                while lo < hi and nums[lo] >= pivot:
                    lo += 1
                nums[hi] = nums[lo]
            nums[lo] = pivot
            return lo

        lo = 0
        hi = len(nums) - 1
        while True:
            mid = paration(lo, hi)
            if mid == k - 1:
                return nums[mid]
            elif mid > k - 1:
                hi = mid - 1
            else:
                lo = mid + 1


if __name__ == "__main__":
    s = Solution()
    n = [3,2,3,1,2,4,5,5,6]
    k = 4
    print(s.findKthLargest(n, k))
